File: scripts/download-neutralino.ts
import { unzipSync } from "fflate";
import { dirname } from "@std/path";
import { getNeutralinoBinaryName } from "../main.ts";

const NEUTRALINO_VERSION = "6.7.0";

const binName = getNeutralinoBinaryName(Deno.build.os, Deno.build.arch);
const destDir = `${import.meta.dirname!}/..`;
const binPath = `${destDir}/${binName}`;


const zipUrl = `https://github.com/neutralinojs/neutralinojs/releases/download/v${NEUTRALINO_VERSION}/neutralinojs-v${NEUTRALINO_VERSION}.zip`;

console.log(`Downloading Neutralinojs v${NEUTRALINO_VERSION}...`);

const response = await fetch(zipUrl);
if (!response.ok) {
  console.error(`Failed to download: HTTP ${response.status}`);
  Deno.exit(1);
}

const zipData = new Uint8Array(await response.arrayBuffer());
const files = unzipSync(zipData);

for (const [name, data] of Object.entries(files)) {
  const outPath = `${destDir}/${name}`;
  await Deno.mkdir(dirname(outPath), { recursive: true });
  await Deno.writeFile(outPath, data);
}

if (Deno.build.os !== "windows") {
  await Deno.chmod(binPath, 0o755);
}

console.log(`Downloaded: ${Object.keys(files).join(", ")}`);

File: src/bin/download-neutralino.rs
use markup_img::{NEUTRALINO_VERSION, get_neutralino_binary_name, is_windows_executable};
use std::fs::{self, File};
use std::io;
use std::path::PathBuf;
use zip::ZipArchive;

fn dest_dir() -> Result<PathBuf, Box<dyn std::error::Error>> {
    Ok(markup_img::find_runtime_root()?)
}

#[cfg(unix)]
fn set_executable_if_needed(path: &std::path::Path) -> io::Result<()> {
    use std::os::unix::fs::PermissionsExt;

    if is_windows_executable(path) {
        return Ok(());
    }

    let mut perms = fs::metadata(path)?.permissions();
    perms.set_mode(0o755);
    fs::set_permissions(path, perms)
}

#[cfg(not(unix))]
fn set_executable_if_needed(_path: &std::path::Path) -> io::Result<()> {
    Ok(())
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let bin_name = get_neutralino_binary_name(std::env::consts::OS, std::env::consts::ARCH)?;
    let root = dest_dir()?;
    let zip_url = format!(
        "https://github.com/neutralinojs/neutralinojs/releases/download/v{0}/neutralinojs-v{0}.zip",
        NEUTRALINO_VERSION
    );

    println!("Downloading Neutralinojs v{NEUTRALINO_VERSION}...");
    let mut response = ureq::get(&zip_url).call()?;
    let bytes = response.body_mut().read_to_vec()?;
    let reader = std::io::Cursor::new(bytes);
    let mut archive = ZipArchive::new(reader)?;
    let mut names = Vec::new();

    for index in 0..archive.len() {
        let mut file = archive.by_index(index)?;
        let out_path = root.join(file.name());

        if file.name().ends_with('/') {
            fs::create_dir_all(&out_path)?;
            continue;
        }

        if let Some(parent) = out_path.parent() {
            fs::create_dir_all(parent)?;
        }

        let mut output = File::create(&out_path)?;
        io::copy(&mut file, &mut output)?;
        set_executable_if_needed(&out_path)?;
        names.push(file.name().to_string());
    }

    println!("Downloaded: {}", names.join(", "));
    let _ = root.join(bin_name);
    Ok(())
}
